src/Lean/Elab/Syntax.lean

```diff
@@ -191,26 +191,31 @@ where
     return (stx, stackSz)
 
   processNullaryOrCat (stx : Syntax) := do
-    match (← elabParserName? stx[0]) with
+    let ident := stx[0]
+    let id := ident.getId.eraseMacroScopes
+    -- run when parser is neither a decl nor a cat
+    let default := do
+      if (← Parser.isParserAlias id) then
+        ensureNoPrec stx
+        return (← processAlias ident #[])
+      throwError "unknown parser declaration/category/alias '{id}'"
+    match (← elabParserName? ident) with
     | some (.parser c (isDescr := true)) =>
       ensureNoPrec stx
       -- `syntax _ :=` at least enforces this
       let stackSz := 1
       return (mkIdentFrom stx c, stackSz)
     | some (.parser c (isDescr := false)) =>
+      if (← Parser.getParserAliasInfo id).declName == c then
+        -- prefer parser alias over base declaration because it has more metadata, #2249
```

Comment:
I decided to restrict the change to the specific issue at hand where preferring the declaration is never useful. Looking at https://github.com/leanprover/lean4/blob/master/tests/lean/run/parserAliasShadow.lean, it didn't appear to be the correct choice in general anymore.